src/main.cpp

```diff
@@ -3557,6 +3597,26 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
                                block.vtx[0].GetValueOut(), blockReward),
                                REJECT_INVALID, "bad-cb-amount");
 
+    // Ensure that the total chain supply is consistent with the value in each pool.
+    if (!fJustCheck &&
+            pindex->nChainTotalSupply &&
+            pindex->nTransparentValue &&
+            pindex->nChainSproutValue &&
+            pindex->nChainSaplingValue &&
+            pindex->nChainOrchardValue)
+    {
+        auto expectedChainSupply =
+            pindex->nTransparentValue.value() +
+            pindex->nChainSproutValue.value() +
+            pindex->nChainSaplingValue.value() +
+            pindex->nChainOrchardValue.value();
+        if (expectedChainSupply != pindex->nChainTotalSupply.value()) {
+            // This may be added as a rule to ZIP 209 and return a failure in a future soft-fork.
+            error("ConnectBlock(): chain total supply (%d) does not match sum of pool balances (%d) at height %d",
+                    pindex->nChainTotalSupply.value(), expectedChainSupply, pindex->nHeight);
```

Comment:
The meagre should include all the individual pool balances.

Reply:
@str4d and had considered that but actively decided against it in our pairing, since one can just look it up with `getblock`.